34-cellular-modbus-client/README.md

```diff
@@ -0,0 +1,427 @@
+
+##### Reads
+
+On success, reading coils or discrete inputs has this response format:
+
+```json
+{
+    "bits": [250, 112],
+    "seq_num": 0
+}
+```
+
+`bits` is a byte array packed in the same fashion as described in [Write Multiple Coils](#write-multiple-coils).
+
+Successful reading of registers has this response format:
+
+```json
+{
+    "regs": [291, 43981],
+    "seq_num": 5
+}
+```
+
+`regs` is an array of 16-bit register values.
+
+##### Writes
+
+All successful write operations have the same response format:
+
+```json
+{
+    "seq_num": 1
+}
+```
+
+##### Errors
+
+If any request results in an error, a response of this form will be added to `responses.qo`:
+
+```json
+{
+    "error": "Illegal data address",
+    "seq_num": 11
+}
+```
+
+`error` is a string describing the error. In this example, a bad address was supplied in the `addr` field of the corresponding request.
+
+### Building and Flashing
+
+To build and upload the firmware onto the MCU, you'll need VS Code with the PlatformIO extension.
+
+1. Download and install [Visual Studio Code](https://code.visualstudio.com/).
+1. Install the [PlatformIO IDE extension](https://marketplace.visualstudio.com/items?itemName=platformio.platformio-ide) via the Extensions menu of Visual Studio Code.
+1. Click the PlatformIO icon on the left side of VS Code, then click Pick a folder, and select the the firmware directory, `34-cellular-modbus-client/firmware`.
+1. In the file explorer, open `main.cpp` and uncomment this line: `// #define PRODUCT_UID "com.my-company.my-name:my-project"`. Replace `com.my-company.my-name:my-project` with the [ProductUID of the Notehub project](https://dev.blues.io/notehub/notehub-walkthrough/#finding-a-productuid) you created in [Notehub Setup](#notehub-setup).
+1. Click the PlatformIO icon again, and under the Project Tasks menu, click Build to build the firmware image.
+1. Under the Project Tasks menu, click Upload to upload the firmware image to the MCU.
+
+From here, you can view logs from the firmware over serial with a terminal emulator (e.g. minicom). On Linux, the serial device will be something like `/dev/ttyACM0`. Use a baud rate of 115200 and [8-N-1](https://en.wikipedia.org/wiki/8-N-1) for the serial parameters.
+
+## Testing
+
+The tests described in this section rely on the server.py script. This script uses [pymodbus](https://github.com/pymodbus-dev/pymodbus) to run a Modbus server on your development PC. It's based on pymodbus's [datastore_simulator.py example](https://github.com/pymodbus-dev/pymodbus/blob/dev/examples/datastore_simulator.py). You can check out server.py's `config` dictionary to see how the coils and registers are addressed. Refer to [pymodbus's documentation](https://github.com/pymodbus-dev/pymodbus/blob/dev/doc/source/library/simulator/config.rst) for details on how the memory layout of the server is configured.
```

Comment:
```suggestion
The tests described in this section rely on the `server.py` script. This script uses [`pymodbus`](https://github.com/pymodbus-dev/pymodbus) to run a Modbus server on your development PC. It's based on pymodbus's [datastore_simulator.py example](https://github.com/pymodbus-dev/pymodbus/blob/dev/examples/datastore_simulator.py). You can check out server.py's `config` dictionary to see how the coils and registers are addressed. Refer to [pymodbus's documentation](https://github.com/pymodbus-dev/pymodbus/blob/dev/doc/source/library/simulator/config.rst) for details on how the memory layout of the server is configured.
```

Reply:
What Mat said, and there are a few other mentions of `server.py` below that should also be put in code tags for consistency.

I’d also make the first mention `server.py` a link to the file, because it’s not obvious you’re referring to a file in this repo (I had to check).